Export/vp 364 catalog category export

akak1977 opened the pull request.

akak1977 committed:
VP-364 export full catalog data

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+            return Items.Any();
+        }
+
+        private void CalculateCounts()
```
The idea of this method is to have all states TotalCount filled, not just to get overall TotalCount:
https://github.com/VirtoCommerce/vc-platform-core/pull/521/files#diff-025b92c5e3684f6c6184640d95e9ba6dR120.
Thus we can use these counts in future calculation which will simplify paging logic

akak1977 replied:
There is the same, because each of FetchFunc fill TotalCount for state

yecli replied:
Really, I see this now.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+{
+    public class CatalogFullExportPagedDataSource : IPagedDataSource
+    {
+        private class FetchResult
```
Do we need really need this class? Seems `ExportDataSourceState` could incapsulate its functionality.

akak1977 replied:
OK.Removed

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+                FetchFunc = fetchFunc;
+            }
+
+            public void SetFetchResult<T>(GenericSearchResult<T> searchResult) where T : Entity
```
Same for this - maybe its redundant. For inner class I even do not see necessity in having constructor for properties filling.

akak1977 replied:
Removed

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+
+        private bool _statesInitialized = false;
+
+        private readonly ExportDataSourceState[] _exportDataSourceStates = new ExportDataSourceState[5];
```
Could you use array for this? Initialization will become easier.

akak1977 replied:
Removed

akak1977 replied:
Here is an array already. Let's discuss on monday.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.FetchResult.Result = Array.Empty<IExportable>();
+                var overlap = skip - curStateStartPos; // Positive overlap should fall into skip, negative into take
```
This logic seems harder to me than when we have TotalCounts already filled like here:
https://github.com/VirtoCommerce/vc-platform-core/pull/521/files#diff-025b92c5e3684f6c6184640d95e9ba6dR89

Could we replace it (with breaking the cycle when portion is fully filled)?

akak1977 replied:
I don't understand. There TotalCounts filled too. Need to discuss.

yecli replied:
Yes, lets discuss on this.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+
+        private void InitExportDataSourceStates()
+        {
+            if (!_statesInitialized)
```
Why do we need this one? It is non shared class with this method called one inside constructor.

akak1977 replied:
Removed. Incapsulated into DataQuery setter.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,211 @@
+        private class ExportDataSourceState
+        {
+            public FetchResult FetchResult;
+            public SearchCriteriaBase SearchCriteria;
```
Maybe we could make generic class with and use ATP to instantiate it on creation to exclude from constructor:
```
class ExportDataSourceState<TSearchCriteria> where TSearchCriteria: SearchCriteriaBase
{
public SearchCriteriaBase SearchCriteria = AbstractTypeFactory<TSearchCriteria>.TryCreateInstance();
}
```

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Test/VirtoCommerce.CatalogModule.Test.csproj:
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Please add tests for pagination here

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/ExportableCatalog.cs:
```diff
@@ -0,0 +1,27 @@
+using System.Linq;
```
Need to remove?

akak1977 replied:
No :-)

yecli replied:
Where this class is used? I cannot find the place, that's why i think we could remove it.

akak1977 replied:
FRomModel:
Languages?.Select

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Web/Module.cs:
```diff
@@ -169,9 +169,19 @@ public void Initialize(IServiceCollection serviceCollection)
             });
 
             #endregion
+
```
Please rebase on the latest export/dev and apply new factory generation scheme.

akak1977 replied:
OK. Done

yecli replied:
In new scheme registrations like below are not needed, you just need to register factory

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -102,7 +102,7 @@ public CatalogFullExportDataQuery DataQuery
                 _exportDataSourceStates.Add(new ExportDataSourceState
                 {
-                    SearchCriteria = new ProductSearchCriteria() { ResponseGroup = (ItemResponseGroup.Full & ~ItemResponseGroup.Variations).ToString() },
+                    SearchCriteria = new ProductSearchCriteria() { CatalogIds = _dataQuery.CatalogIds, ResponseGroup = (ItemResponseGroup.Full & ~ItemResponseGroup.Variations).ToString() },
```
Are variations included in search result? I think we need to set
```
SearchInVariations = true
```
for that.

akak1977 replied:
I'm not sure. This initialization just copied from old export.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,182 @@
+
+            Task.WhenAll(taskList).GetAwaiter().GetResult(); // Ensures completing all fetches
+
+            ((List<IExportable>)Items).AddRange(_exportDataSourceStates.SelectMany(x => x.Result));
```
Why do we need this cast?

akak1977 replied:
Ok. I'll transform it to Items.Concat...

akak1977 replied:
Ok. Removed.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs:
```diff
@@ -0,0 +1,182 @@
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
```
If name the variable properly there is no need in such comments.

yecli replied:
Also I think we even do not need current position in state. As we always have Skip and Take, and state TotalCount's. These data are sufficient to take portion of data.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Test/ExportImport2Test.cs:
```diff
@@ -0,0 +1,91 @@
+        /// </summary>
+        /// <returns></returns>
+        [Fact]
+        public Task PagingThruDataSourcesTest()
```
Please rename to `CatalogFullExportPagedDataSource_Paging_Valid`

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Test/ExportImport2Test.cs:
```diff
@@ -0,0 +1,91 @@
+                fetchResult.AddRange(catalogDataSource.Items);
+            }
+
+            Assert.Equal("1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22", string.Join(',', fetchResult.Select(x => x.Id)));
```
Let's compare each assert.
This assert does not check count of pages and its content, only flow.

yecli commented on Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Web/Module.cs:
```diff
@@ -231,7 +243,15 @@ public void PostInitialize(IApplicationBuilder appBuilder)
+                    );
+
             AbstractTypeFactory<ExportDataQuery>.RegisterType<ProductExportDataQuery>();
+            AbstractTypeFactory<ExportDataQuery>.RegisterType<CatalogFullExportDataQuery>();
```
No need in these registrations too, they are done in ExportedTypeDefinitionBuilder now.

yecli replied:
As all the changes below.

akak1977 replied:
OK

yecli commented on Modules/vc-module-export/VirtoCommerce.ExportModule.Core/Model/ExportedTypeDefinition.cs:
```diff
@@ -42,6 +42,6 @@ public class ExportedTypeDefinition
         /// </summary>
         [JsonIgnore]
-        public Func<ExportDataQuery, IPagedDataSource> ExportedDataSourceFactory { get; set; }
+        public IPagedDataSourceFactory DataSourceFactory { get; set; }
```
Why these are the changes? I think it is a part of export/dev already. Did you rebase on this branch?

akak1977 replied:
Yes. rebased. Agree, it looks strange. May be this is a something due conflict changes.

akak1977 committed:
Fetch: Skip iterating thru next states after fullfilling current take.
Review-based changes.

akak1977 committed:
Metafields added to filter by Catalogs

yecli committed:
VP-364: Refactored CatalogFullExportPagedDataSource;
Created ComplexExportPagedDataSource in Export module;
Moved ComplexExportPagedDataSource tests from Catalog module to Export;
